File: programmers/java/programmers220726/programmers1.java
package programmers.java.programmers220726;

import java.util.*;

// https://school.programmers.co.kr/learn/courses/30/lessons/67256
public class programmers1 {
    class Pair {
        private int x;
        private int y;

        public Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public int getX() {
            return x;
        }

        public int getY() {
            return y;
        }
    }

    class Solution {
        private Map<Integer, Pair> indexMap = new HashMap<>();
        private Set<Integer> left_num = Set.of(1, 4, 7);
        private Set<Integer> right_num = Set.of(3, 6, 9);

        private Pair leftHand = new Pair(3, 0);
        private Pair rightHand = new Pair(3, 2);

        {
            indexMap.put(1, new Pair(0, 0));
            indexMap.put(2, new Pair(0, 1));
            indexMap.put(3, new Pair(0, 2));
            indexMap.put(4, new Pair(1, 0));
            indexMap.put(5, new Pair(1, 1));
            indexMap.put(6, new Pair(1, 2));
            indexMap.put(7, new Pair(2, 0));
            indexMap.put(8, new Pair(2, 1));
            indexMap.put(9, new Pair(2, 2));
            indexMap.put(0, new Pair(3, 1));
        }

        public String solution(int[] numbers, String hand) {
            String answer = "";

            for (int num : numbers) {
                if (left_num.contains(num)) {
                    leftHand = indexMap.get(num);
                    answer += "L";
                    continue;
                }
                if (right_num.contains(num)) {
                    rightHand = indexMap.get(num);
                    answer += "R";
                    continue;
                }
                answer += move(num, hand);
            }

            return answer;
        }

        private String move(int target, String hand) {
            Pair target_location = indexMap.get(target);

            int left_distance = distance(leftHand, target_location);
            int right_distance = distance(rightHand, target_location);

            if (left_distance > right_distance) {
                rightHand = target_location;
                return "R";
            }
            if (right_distance > left_distance) {
                leftHand = target_location;
                return "L";
            }

            if ("right".equals(hand)) {
                rightHand = target_location;
                return "R";
            } else {
                leftHand = target_location;
                return "L";
            }
        }

        private int distance(Pair a, Pair b) {
            return Math.abs(a.getX() - b.getX()) + Math.abs(a.getY() - b.getY());
        }
    }
}

File: programmers/python/programmers220814/programmers1.py
// https://school.programmers.co.kr/learn/courses/30/lessons/77485
def solution(rows, columns, queries):
    arr = [[row*columns + col + 1 for col in range (columns)] for row in range (rows)]
    answer = []

    for x1, y1, x2, y2 in queries :
        target, arr = rotate(x1-1, y1-1, x2-1, y2-1, arr)
        answer.append(target)

    return answer

def rotate(x1, y1, x2, y2, arr) :
    target = arr[x1][y1] ## 시작
    tmp = arr[x1][y1]

    ## 왼쪽
    for x in range (x1, x2) :
        arr[x][y1] = arr[x+1][y1]
        target = min(target, arr[x+1][y1])

    ## 아래쪽
    for y in range (y1, y2) :
        arr[x2][y] = arr[x2][y+1]
        target = min(target, arr[x2][y+1])

    ## 오른쪽
    for x in range (x2, x1, -1) :
        arr[x][y2] = arr[x-1][y2]
        target = min(target, arr[x-1][y2])

    ## 위쪽
    for y in range (y2, y1, -1) :
        if (y-1 == y1) :
            arr[x1][y] = tmp
            target = min(target, tmp)
        else :
            arr[x1][y] = arr[x1][y-1]
            target = min(target, arr[x1][y-1])
    return target, arr